### PR TITLE
e1000: Fix hardcoded BDF

### LINKED ISSUE #51829: qemu_x86: upgrading to q35 breaks networking samples.
**Describe the bug**
with this commit, samples/net/socket/echo_server fails to run
```
commit a2aa462f7d099bf30c1e054a1f38d34359d83418 (refs/bisect/bad)
Author: Andrei Emeltchenko <andrei.emeltchenko@intel.com>
Date:   Thu Oct 27 18:54:13 2022 +0300

    qemu_x86: Use Qemu Q35 machine feature (ICH9)

    Using very old machine does not make sense anymore. Switch to the new
    Q35 (https://wiki.qemu.org/Features/Q35). Among other things it allows
    to test SMBus and watchdog.

    Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

diff --git a/boards/x86/qemu_x86/board.cmake b/boards/x86/qemu_x86/board.cmake
index b4f7c03639..934b078154 100644
--- a/boards/x86/qemu_x86/board.cmake
+++ b/boards/x86/qemu_x86/board.cmake
@@ -60,6 +60,7 @@ endif()
 set(QEMU_FLAGS_${ARCH}
   -m ${QEMU_MEMORY_SIZE_MB}
   -cpu ${QEMU_CPU_TYPE_${ARCH}}${QEMU_CPU_FLAGS}
+  -machine q35
   -device isa-debug-exit,iobase=0xf4,iosize=0x04
   ${REBOOT_FLAG}
   -nographic
```

**To Reproduce**
1. mkdir build; cd build
2. west build -b qemu_x86 -- -DCONF_FILE=prj.conf -DOVERLAY_CONFIG="overlay-e1000.conf"
3. west build -t run
```
-- west build: running target run
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: qemu32,+nx,+pae
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)


iPXE (http://ipxe.org) 00:02.0 CA00 PCI2.10 PnP PMM+00392120+002F2120 CA00



Booting from ROM..
FAILED: zephyr/CMakeFiles/run_qemu /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/net/sockets/echo_server/build/zephyr/CMakeFiles/run_qemu
cd /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/net/sockets/echo_server/build && /home/ubuntu/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/qemu-system-i386 -m 4 -cpu qemu32,+nx,+pae -machine q35 -device isa-debug-exit,iobase=0xf4,iosize=0x04 -no-reboot -nographic -no-acpi -nic tap,model=e1000,script=no,downscript=no,ifname=zeth -pidfile qemu.pid -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -kernel /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/net/sockets/echo_server/build/zephyr/zephyr.elf
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /home/shared/disk/zephyr_project/zephyr_test/zephyr/samples/net/sockets/echo_server/build --target run

```

**Expected behavior**
the qemu_x86 can start successfully

**Impact**
showstopper

**Environment (please complete the following information):**
 - OS: (e.g. Linux, , )
 - Toolchain (e.g Zephyr SDK, )
 - Commit SHA or Version used: a2aa462f7d099bf30c1e054a1f38d34359d83418 

**Additional context**
this impact the maxpro networking tcpip testing, as we testing tcpip stack on qemu_x86 platforms

### PR DESCRIPTION
**Pull request by finikorg**
Correct e1000 hardcoded BDF, making it possible to use the same driver when Qemu parameters change.

Fixes #51829

### CODE REVIEW

**Comment by jhedberg** [comment]
Has someone looked into how the tests should be updated to pass with q35?

**Comment by henrikbrixandersen** [comment]
> Has someone looked into how the tests should be updated to pass with q35?

As far as I can see, it is just a matter of where the emulated E1000 is placed on the PCIe bus. Changing `PCIE_BDF(0, 3, 0)` to `PCIE_BDF(0, 2, 0)` should make it work.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
This will result in different PCIe layouts for `qemu_x86_64` and `qemu_x86`. I suggest just updating the E1000 driver to use the new BDF (or even better, make the driver obtain the BDF from dts).

**Comment by henrikbrixandersen** [comment]
I am not talking about PCIe IRQ allocation. I simple refer to https://github.com/zephyrproject-rtos/zephyr/blob/7756809c972d5fff0b52d85cfda5f6f20628a2c2/drivers/ethernet/eth_e1000.c#L238 which hardcodes the BDF for the E1000 driver.

**Comment by finikorg** [comment]
> I am not talking about PCIe IRQ allocation. I simple refer to
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/7756809c972d5fff0b52d85cfda5f6f20628a2c2/drivers/ethernet/eth_e1000.c#L238
> which hardcodes the BDF for the E1000 driver.

This can be done. I was thinking to rewrite e1000 using PCIE IRQ allocation.

**Comment by henrikbrixandersen** [comment]
> This can be done. I was thinking to rewrite e1000 using PCIE IRQ allocation.

That would be nice, but would require dynamic PCIe IRQ allocation supported on non-64 bit x86. I skipped dynamic IRQ allocation for this reason in #51750.

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Review by enjiamai — DISMISSED** [review]

File: drivers/ethernet/eth_e1000.c [hunk 1]
```diff
@@ -230,18 +230,16 @@ static void e1000_isr(const struct device *ddev)
...
 }
 
-#define PCI_VENDOR_ID_INTEL	0x8086
-#define PCI_DEVICE_ID_I82540EM	0x100e
 
 int e1000_probe(const struct device *ddev)
 {
-	const pcie_bdf_t bdf = PCIE_BDF(0, 3, 0);
+	/* PCI ID is decoded into REG_SIZE */
+	pcie_bdf_t bdf = pcie_bdf_lookup(DT_INST_REG_SIZE(0));
```
[the remark this answers is not in the document]
Comment (finikorg) [reply drivers/ethernet/eth_e1000.c:237]: No, as before

**Comment by jhedberg** [comment]
Is the BDF still correct after the CAN support for qemu that was just merged?

**Review by tbursztyka — DISMISSED** [review]

**Comment by finikorg** [comment]
> Is the BDF still correct after the CAN support for qemu that was just merged?

Should not matter now. [540fecb](https://github.com/zephyrproject-rtos/zephyr/pull/51845/commits/540fecb4f84db7f04a981bf3bd449c3f7eba8801) I think BDF might change depending on Qemu command line options.

**Comment by finikorg** [comment]
> > Is the BDF still correct after the CAN support for qemu that was just merged?
> 
> Should not matter now. [540fecb](https://github.com/zephyrproject-rtos/zephyr/pull/51845/commits/540fecb4f84db7f04a981bf3bd449c3f7eba8801) I think BDF might change depending on Qemu command line options.

I will rebase since pcie gets duplicated

**Review by henrikbrixandersen — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**